### PR TITLE
WIP default_values

### PR DESCRIPTION
**Pull request by benjello**
This a a non working tentative which tries to reimplement @AlexisEidelman default_value declaration
@gdementen if you have a bit of time to have look at it I would be very glad.
And if you think, that it I am in the wrong direction, please be blunt !
Thanks

### CODE REVIEW

File: liam2/data.py [hunk 1]
```diff
@@ -193,6 +200,7 @@ def from_table(cls, table, start=0, stop=None, buffersize=10 * 2 ** 20):
         max_buffer_rows = buffersize // dtype.itemsize
         numlines = stop - start
         ca = cls.empty(numlines, dtype)
+        ca.default_values = default_values
```
Comment (gdementen) [inline liam2/data.py]: empty should accept an optional arg, and this line should use it instead

File: liam2/entities.py [hunk 2]
```diff
@@ -131,6 +132,7 @@ def __init__(self, name, fields=None, links=None, macro_strings=None,
             def fdef2field(name, fielddef):
                 initialdata = True
                 output = True
+                default_value = -1
```
Comment (gdementen) [inline liam2/entities.py]: the default default value should depend on type (nan for float, False for bool)

File: liam2/entities.py [hunk 3]
```diff
@@ -138,13 +140,14 @@ def fdef2field(name, fielddef):
                         strtype = fielddef['type']
                         initialdata = fielddef.get('initialdata', True)
                         output = fielddef.get('output', True)
+                        default_value = fielddef.get('default', -1)
```
Comment (gdementen) [inline liam2/entities.py]: idem

File: liam2/tests/functional/simulation.yml [hunk 4]
```diff
@@ -2473,6 +2475,13 @@ entities:
                       m_id=-1,
                       hh_id=-1)
 
+            test_default_value:
+                - show("test default value")
+                - qshow(defaulted_variable)
+                - qshow(undefaulted_variable)
+                - assertTrue(all(defaulted_variable == 99))
+                - assertFalse(all(undefaulted_variable == 99))
```
Comment (gdementen) [inline liam2/tests/functional/simulation.yml]: replace this by assertEqual(xxx, -1)? (not sure it works)
or at least assertTrue(all(undefaulted_variable == -1))

Comment (gdementen) [inline liam2/tests/functional/simulation.yml]: you should also add tests for other types

File: liam2/data.py [hunk 5]
```diff
@@ -515,8 +530,8 @@ def copy_table(input_table, output_node, output_dtype=None,
 # this is a special case though because:
 # 1) all arrays have the same columns
 # 2) we have id_to_rownum already computed for each array
-def build_period_array(input_table, output_fields, input_rows, input_index,
-                       start_period):
+def build_period_array(input_table, output_fields, input_rows,
+                    input_index, start_period, default_values={}):
```
Comment (gdementen) [inline liam2/data.py]: indent is wrong

File: liam2/data.py [hunk 6]
```diff
@@ -476,10 +489,12 @@ def copy_chunk(chunk_idx, chunk_num):
         if output_fields is not None:
             # use our pre-allocated buffer (except for the last chunk)
             if len(input_data) == len(expanded_data):
+                default_values = {}
```
Comment (gdementen) [inline liam2/data.py]: this is a bit silly... default_values should be an argument of append_table

**Comment by gdementen** [comment]
There are strange things in this proposition, but it's hard to tell what exactly without thinking hard about it and I am too tired to do it today and will probably not have time to do it for a few weeks.

**Comment by benjello** [comment]
Thanks @gdementen, it is messy because it is mainly a merge with a few fix. I will try on my own for a moment.

**Comment by benjello** [comment]
@gdementen : should I modify the get_missing_* by adding a default_values argument or should I add it only to ColumnArray.empty and the calls to this method in ColumnArray.from_table and ColumnArray.from_records ?

**Comment by gdementen** [comment]
@benjello I don't know. You should see what works best. I would find it elegant if we could suppress the concept of missing value, and only have default values which themselves default to the current values we use for missing values depending on the type of field.

**Comment by benjello** [comment]
It is somehow what we have done in openfisca and it is actually useful. I
will try but I may come with more questions ;-) so stop when you get bored

Le jeu. 29 oct. 2015 09:03, Gaëtan de Menten <notifications@github.com> a
écrit :

**Comment by gdementen** [comment]
I won't get bored, but I am low on time... On the other hand I want to avoid slowing you down or discouraging you from contributing to LIAM2.

**Comment by benjello** [comment]
We are all low on time ;-)

Le jeu. 29 oct. 2015 09:17, Gaëtan de Menten <notifications@github.com> a
écrit :

**Comment by gdementen** [comment]
FWIW, np.full is numpy1.8+, so we'd need to bump the requirements in INSTALL.

**Comment by benjello** [comment]
I introduced default_values everywhere the was a missing_* but still it the test do not pass. I confess I do not figure the entire process very well so i could use a hint about how to debug this kind of problem ;-)

**Comment by gdementen** [comment]
The line "defaulted_integer_variable: [99 99 99 ..., -1 -1 -1] done (0 ms elapsed)." makes me suspect the new borns do not use the default value, which means the thing is probably working as it should for the initial population but not for newborns. Did you implement it?

**Comment by benjello** [comment]
Good catch @gdementen. Tests do pass now. I can add more tests if needed. 
You should carefully review this PR, i was not always very confident about what i was modifying.
I still lack a clear vision of the global architecture.

**Comment by gdementen** [comment]
I will not have time for a in-depth review & merge for a couple of weeks, but in the mean time, could you;
* provide documentation 
* change the default value for the default arguments to None instead of {} (and update the code accordingly). Using a mutable default value for an argument is asking for trouble.
* ideally add a mention in the changelog at doc/usersguide/source/changes/version_0_11.rst.inc
Thanks!

**Comment by benjello** [comment]
I wil try to find some time for that.

File: doc/usersguide/source/model.rst [hunk 7]
```diff
@@ -219,6 +219,10 @@ However, in practice, there are often some fields which are not present in the
 input file. They will need to be calculated later by the model, and you need to
 tell LIAM2 that the field is missing, by using `initialdata: False` in the
 definition for that field (see the *agegroup* variable in the example below).
+The fields that are not present in the initial file can also be initialized to
+a specific value by using the `default: some_default_value`. If not present they
+are initialized to the default of the field type which are False for boolean and
+0 for integer and float (see the *alive* variable in the example below).
```
Comment (gdementen) [inline doc/usersguide/source/model.rst]: this is not True (or at least was not True). -1 for integer and 'nan' for float

File: doc/usersguide/source/model.rst [hunk 8]
```diff
@@ -239,14 +243,14 @@ functions but not stored in the output file.
                 - age:      int
                 - agegroup: {type: int, initialdata: False}
                 - temporary: {type: int, output: False}
-
+                - alive: {type: int, initialdata: True}
```
Comment (gdementen) [inline doc/usersguide/source/model.rst]: I guess you meant {type: int, default: True}

Comment (benjello) [inline doc/usersguide/source/model.rst]: Sorry I will fix that.

File: doc/usersguide/source/model.rst (hunk 7 above)
Comment (gdementen) [inline doc/usersguide/source/model.rst]: You do no speak about new(). In fact I think a separate paragraph and saying it applies to both new() and fields not present in the input bla bla... would be better, what do you think?

Comment (benjello) [inline doc/usersguide/source/model.rst]: You are right. Will do that.

**Comment by gdementen** [comment]
Merged! Thanks a lot!

**Comment by benjello** [comment]
I hope you checked it carefully since I was not always confortable with what I was modifying.

**Comment by gdementen** [comment]
Don't worry, I did not merge it as-is. I did check & modify it. There is still a bug left, but I wanted to add a test case for it before fixing it.

**Pull request closed without merging**